server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java

```diff
@@ -1271,21 +1285,50 @@ public void testNodeCannotJoinIfJoinValidationFailsOnJoiningNode() {
         }
     }
 
-    public void testClusterCannotFormWithFailingJoinValidation() {
+    public void testClusterCannotFormWithFailingJoinValidation() throws Exception {
+        clusterCannotFormWithFailingJoinValidation(false);
+    }
+
+    /**
+     * Forms a random sized cluster and then disables join validation on either a random majority subset or all cluster nodes. Then checks
+     * that election fails.
+     *
+     * @param failJoinOnAllNodes this controls whether to fail join on all nodes or only a majority subset. The atomic register CAS election
+     *                           strategy will succeed in electing a master if any node can vote (even the master candidate voting for
+     *                           itself).
+     * @throws Exception
+     */
+    protected void clusterCannotFormWithFailingJoinValidation(boolean failJoinOnAllNodes) throws Exception {
         try (Cluster cluster = new Cluster(randomIntBetween(1, 5))) {
-            // fail join validation on a majority of nodes in the initial configuration
-            randomValueOtherThanMany(
-                nodes -> cluster.initialConfiguration.hasQuorum(
-                    nodes.stream().map(ClusterNode::getLocalNode).map(DiscoveryNode::getId).collect(Collectors.toSet())
-                ) == false,
-                () -> randomSubsetOf(cluster.clusterNodes)
-            ).forEach(cn -> cn.extraJoinValidators.add((discoveryNode, clusterState) -> {
+            List<ClusterNode> clusterNodesToFailJoin;
+            if (failJoinOnAllNodes) {
+                // The AtomicRegister strategy succeeds if a master candidate votes for itself, so we must disable all nodes from voting so
+                // that none of them can self-elect.
+                clusterNodesToFailJoin = cluster.clusterNodes;
+            } else {
+                // Fetch a random subset of cluster nodes that form a quorum (majority subset).
+                clusterNodesToFailJoin = randomValueOtherThanMany(
+                    nodes -> cluster.initialConfiguration.hasQuorum(
+                        nodes.stream().map(ClusterNode::getLocalNode).map(DiscoveryNode::getId).collect(Collectors.toSet())
+                    ) == false,
+                    () -> randomSubsetOf(cluster.clusterNodes)
+                );
+            }
+
+            // Fail join validation on the set of nodes so that election will fail in the initial configuration.
+            clusterNodesToFailJoin.forEach(cn -> cn.extraJoinValidators.add((discoveryNode, clusterState) -> {
                 throw new IllegalArgumentException("join validation failed");
             }));
+
             cluster.bootstrapIfNecessary();
+
+            // Run the cluster for 10 seconds to give the cluster some time to elect a master.
+            // It's possible stabilisation takes longer, but very unlikely.
```

Comment:
```suggestion
            // It's possible stabilisation takes longer, but essentially impossible that it _always_ takes longer.
```

Reply:
Yes, even if it only stabilized quicker than 10s only 0.1% of the time that'd still be enough for CI to pick it up pretty fast, and even one-in-a-million failures get noticed eventually. And in practice I think it stabilizes quicker than 10s way more often than 0.1% of the time.

Reply:
IOW "very unlikely" isn't the decider here, even if it were very unlikely (but not impossible) this would still be a valid way to write this test.

Reply:
Updated the text 👍

My first thought here was that if we ever do accidentally make election take a long time, then this test will continue to happily succeed (falsely) without complaint. So technically risky, even if unlikely 🤷‍♀️ Presumably/Hopefully other tests would complain if that were to happen.